accepted/cranelift-assembler.md

```diff
@@ -0,0 +1,368 @@
+instruction for `cranelift-codegen`'s x64 backend ([`abrown#assembler-collaborate`])
+
+[`abrown#assembler-collaborate`]: https://github.com/bytecodealliance/wasmtime/compare/main...abrown:wasmtime:assembler-collaborate
+
+### DSL
+[DSL]: #dsl
+
+We describe the assembler instructions using single lines in a domain-specific language (DSL)
+written in Rust. The goal here is brevity and transference from the developer manual &mdash; ideally
+one could look at the manual, look at the DSL, and visually check equivalence.
+
+The Intel [manual] (and others, like Felix Cloutier's excellent HTML [transformation]) rely on an instruction table that describes the instruction name, opcode, feature set, etc.
+
+[manual]: https://www.intel.com/content/www/us/en/developer/articles/technical/intel-sdm.html
+[transformation]: https://www.felixcloutier.com/x86
+
+![Instruction Table](./cranelift-assembler-instruction.png)
+
+The "Op/En" column refers to a secondary table that the manual calls the "Instruction Operand
+Encoding" table but we will refer to as a __format__ table:
+
+![Format (Encoding) Table](./cranelift-assembler-format.png)
+
+Our DSL merges those tables into single lines:
+
+```rust
+inst("andb", fmt("I", [rw(al), r(imm8)]), rex(0x24).ib(), None),
+inst("andw", fmt("I", [rw(ax), r(imm16)]), rex(0x25).prefix(_66).iw(), None),
+inst("andl", fmt("I", [rw(eax), r(imm32)]), rex(0x25).id(), None),
+inst("andq", fmt("I_SXLQ", [rw(rax), sxq(imm32)]), rex(0x25).w().id(), None),
+...
+inst("andb", fmt("RM", [rw(r8), r(rm8)]), rex(0x22).r(), None),
+inst("andb", fmt("RM_SXBQ", [rw(r8), r(rm8)]), rex(0x22).w().r(), None),
+inst("andw", fmt("RM", [rw(r16), r(rm16)]), rex(0x23).prefix(_66).r(), None),
+inst("andl", fmt("RM", [rw(r32), r(rm32)]), rex(0x23).r(), None),
+inst("andq", fmt("RM", [rw(r64), r(rm64)]), rex(0x23).w().r(), None),
+```
+
+Some explanation:
+- an `inst` is built up from an __instruction mnemonic__, a __format__ (`fmt`), an __encoding__
+  (`rex` here), and CPUID __features__ (`None` here)
+- the __format__ describes the instruction operands precisely and has a __format name__
+- the __encoding__ describes how to emit the machine code and eventually would have `vex` and `evex`
+  constructors
+- the __feature__ set will eventually allow one to write expressions like `AVX512_X | AVX512_Y`
+
+Note that this DSL will demand that instruction names be unique: but we used `andb` twice above! To
+solve this, the DSL concatenates the __instruction mnemonic__ and the __format name__. E.g., we
+would refer to the first instruction above as `Inst::andb_i`. Unfortunately, this plan is almost,
+but not quite, sufficient for uniqueness; we manually append descriptors like `SXBQ` where the
+table-provided names are simply not specific enough. Perhaps we could auto-append this suffix in the
+future, but we want it to be quite clear what instruction name will be generated in the next step.
+
+### Generation
```

Comment:
> For example, we should not encode the DSL in toml or yaml and require a new build-dep for parsing that markup language. Nor should we write our own parser for the DSL that has to be built and run from build.rs.

Personally I'd push back on this and say that whether or not this is acceptable is a function of the various other tradeoffs in play. While we should be cognizant of compile time I don't think we should take a policy of "this can't land unless it has 0 compile-time regressions" and instead balance the compile time with the maintainability win (and other motivations) in this RFC

Reply:
Maybe it's not super clear in the text, @fitzgen, but there is no new parsing or markup language involved here. I think I would have preferred a true DSL but, based on your initial recommendation, I just use Rust code in a `meta` crate much like we do in `cranelift-codegen-meta`. So the DSL is made up of Rust helper functions that create the "AST" &mdash; no additional build dependencies required.